File: app/src/main/java/com/zack/enderplan/view/activity/TypeEditActivity.java
package com.zack.enderplan.view.activity;

import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.StringRes;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.zack.enderplan.App;
import com.zack.enderplan.R;
import com.zack.enderplan.injector.component.DaggerTypeEditComponent;
import com.zack.enderplan.injector.module.TypeEditPresenterModule;
import com.zack.enderplan.presenter.TypeEditPresenter;
import com.zack.enderplan.view.dialog.TypeMarkColorPickerDialogFragment;
import com.zack.enderplan.view.dialog.EditorDialogFragment;
import com.zack.enderplan.view.dialog.TypeMarkPatternPickerDialogFragment;
import com.zack.enderplan.view.contract.TypeEditViewContract;
import com.zack.enderplan.model.bean.FormattedType;
import com.zack.enderplan.model.bean.TypeMarkColor;
import com.zack.enderplan.model.bean.TypeMarkPattern;
import com.zack.enderplan.common.Constant;
import com.zack.enderplan.view.widget.CircleColorView;
import com.zack.enderplan.view.widget.ItemView;

import javax.inject.Inject;

import butterknife.BindString;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class TypeEditActivity extends BaseActivity implements TypeEditViewContract {

    @BindView(R.id.toolbar)
    Toolbar mToolbar;
    @BindView(R.id.ic_type_mark)
    CircleColorView mTypeMarkIcon;
    @BindView(R.id.text_type_name)
    TextView mTypeNameText;
    @BindView(R.id.item_type_name)
    ItemView mTypeNameItem;
    @BindView(R.id.item_type_mark_color)
    ItemView mTypeMarkColorItem;
    @BindView(R.id.item_type_mark_pattern)
    ItemView mTypeMarkPatternItem;

    @BindString(R.string.dscpt_unsettled)
    String mUnsettledDscpt;

    @Inject
    TypeEditPresenter mTypeEditPresenter;

    public static void start(Activity activity, int typeListPosition, boolean sharedElementTransition, View sharedElement, String sharedElementName) {
        Intent intent = new Intent(activity, TypeEditActivity.class);
        intent.putExtra(Constant.TYPE_LIST_POSITION, typeListPosition);
        ActivityOptions options;
        if (sharedElementTransition) {
            options = ActivityOptions.makeSceneTransitionAnimation(activity, sharedElement, sharedElementName);
        } else {
            options = ActivityOptions.makeSceneTransitionAnimation(activity);
        }
        activity.startActivity(intent, options.toBundle());
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mTypeEditPresenter.attach();
    }

    @Override
    protected void onInjectPresenter() {
        DaggerTypeEditComponent.builder()
                .typeEditPresenterModule(new TypeEditPresenterModule(this, getIntent().getIntExtra(Constant.TYPE_LIST_POSITION, -1)))
                .appComponent(App.getAppComponent())
                .build()
                .inject(this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mTypeEditPresenter.detach();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    @Override
    public void showInitialView(FormattedType formattedType) {
        setContentView(R.layout.activity_type_edit);
        ButterKnife.bind(this);

        setSupportActionBar(mToolbar);
        setupActionBar();

        onTypeNameChanged(formattedType.getTypeName(), formattedType.getFirstChar());
        onTypeMarkColorChanged(formattedType.getTypeMarkColorInt(), formattedType.getTypeMarkColorName());
        onTypeMarkPatternChanged(formattedType.isHasTypeMarkPattern(), formattedType.getTypeMarkPatternResId(), formattedType.getTypeMarkPatternName());
    }

    @Override
    public void showTypeNameEditorDialog(String originalEditorText) {
        EditorDialogFragment fragment = EditorDialogFragment.newInstance(App.getContext().getString(R.string.title_dialog_type_name_editor), originalEditorText);
        fragment.setOnPositiveButtonClickListener(new EditorDialogFragment.OnPositiveButtonClickListener() {
            @Override
            public void onPositiveButtonClick(String editorText) {
                mTypeEditPresenter.notifyUpdatingTypeName(editorText);
            }
        });
        fragment.show(getSupportFragmentManager(), Constant.TYPE_NAME);
    }

    @Override
    public void onTypeNameChanged(String typeName, String firstChar) {
        mTypeMarkIcon.setInnerText(firstChar);
        mTypeNameText.setText(typeName);
        mTypeNameItem.setDescriptionText(typeName);
    }

    @Override
    public void onTypeMarkColorChanged(int colorInt, String colorName) {
        mTypeMarkIcon.setFillColor(colorInt);
        mTypeMarkColorItem.setDescriptionText(colorName);
    }

    @Override
    public void onTypeMarkPatternChanged(boolean hasPattern, int patternResId, String patternName) {
        mTypeMarkIcon.setInnerIcon(hasPattern ? getDrawable(patternResId) : null);
        mTypeMarkPatternItem.setDescriptionText(hasPattern ? patternName : mUnsettledDscpt);
    }

    @Override
    public void showTypeMarkColorPickerDialog(String defaultColor) {
        TypeMarkColorPickerDialogFragment fragment = TypeMarkColorPickerDialogFragment.newInstance(defaultColor);
        fragment.setOnTypeMarkColorPickedListener(new TypeMarkColorPickerDialogFragment.OnTypeMarkColorPickedListener() {
            @Override
            public void onTypeMarkColorPicked(TypeMarkColor typeMarkColor) {
                mTypeEditPresenter.notifyTypeMarkColorSelected(typeMarkColor);
            }
        });
        fragment.show(getSupportFragmentManager(), Constant.TYPE_MARK_COLOR);
    }

    @Override
    public void showTypeMarkPatternPickerDialog(String defaultPattern) {
        TypeMarkPatternPickerDialogFragment fragment = TypeMarkPatternPickerDialogFragment.newInstance(defaultPattern);
        fragment.setOnTypeMarkPatternPickedListener(new TypeMarkPatternPickerDialogFragment.OnTypeMarkPatternPickedListener() {
            @Override
            public void onTypeMarkPatternPicked(TypeMarkPattern typeMarkPattern) {
                mTypeEditPresenter.notifyTypeMarkPatternSelected(typeMarkPattern);
            }
        });
        fragment.show(getSupportFragmentManager(), Constant.TYPE_MARK_PATTERN);
    }

    @OnClick({R.id.item_type_name, R.id.item_type_mark_color, R.id.item_type_mark_pattern})
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.item_type_name:
                mTypeEditPresenter.notifySettingTypeName();
                break;
            case R.id.item_type_mark_color:
                mTypeEditPresenter.notifySettingTypeMarkColor();
                break;
            case R.id.item_type_mark_pattern:
                mTypeEditPresenter.notifySettingTypeMarkPattern();
                break;
        }
    }

    @Override
    public void showToast(@StringRes int msgResId) {
        Toast.makeText(this, msgResId, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void exit() {
        finish();
    }
}

File: app/src/main/java/net/zackzhang/app/end/view/activity/TypeEditActivity.kt
package net.zackzhang.app.end.view.activity

import android.app.Activity
import android.app.ActivityOptions
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.MenuItem
import android.view.View
import android.view.Window
import butterknife.ButterKnife
import butterknife.OnClick
import kotlinx.android.synthetic.main.activity_type_edit.*
import kotlinx.android.synthetic.main.content_type_edit.*
import net.zackzhang.app.end.App
import net.zackzhang.app.end.R
import net.zackzhang.app.end.common.Constant
import net.zackzhang.app.end.injector.component.DaggerTypeEditComponent
import net.zackzhang.app.end.injector.module.TypeEditPresenterModule
import net.zackzhang.app.end.model.bean.FormattedType
import net.zackzhang.app.end.presenter.TypeEditPresenter
import net.zackzhang.app.end.util.ColorUtil
import net.zackzhang.app.end.view.contract.TypeEditViewContract
import net.zackzhang.app.end.view.dialog.EditorDialogFragment
import net.zackzhang.app.end.view.dialog.TypeMarkColorPickerDialogFragment
import net.zackzhang.app.end.view.dialog.TypeMarkPatternPickerDialogFragment
import javax.inject.Inject

class TypeEditActivity : BaseActivity(), TypeEditViewContract {

    companion object {

        private const val TAG_TYPE_NAME_EDITOR = "type_name_editor"
        private const val TAG_TYPE_MARK_COLOR_PICKER = "type_mark_color_picker"
        private const val TAG_TYPE_MARK_PATTERN_PICKER = "type_mark_pattern_picker"

        fun start(activity: Activity, typeListPosition: Int, enableTransition: Boolean, sharedElement: View, transitionName: String) {
            val intent = Intent(activity, TypeEditActivity::class.java)
            intent.putExtra(Constant.TYPE_LIST_POSITION, typeListPosition)
            intent.putExtra(Constant.ENABLE_TRANSITION, enableTransition)
            if (enableTransition) {
                activity.startActivity(
                        intent,
                        ActivityOptions.makeSceneTransitionAnimation(activity, sharedElement, transitionName).toBundle()
                )
            } else {
                activity.startActivity(intent)
            }
        }
    }

    @Inject
    lateinit var typeEditPresenter: TypeEditPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        typeEditPresenter.attach()
    }

    override fun onInjectPresenter() {
        DaggerTypeEditComponent.builder()
                .typeEditPresenterModule(TypeEditPresenterModule(
                        this,
                        intent.getIntExtra(Constant.TYPE_LIST_POSITION, -1)
                ))
                .appComponent(App.appComponent)
                .build()
                .inject(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        typeEditPresenter.detach()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onAttachFragment(fragment: Fragment) {
        super.onAttachFragment(fragment)

        when (fragment.tag) {
            TAG_TYPE_NAME_EDITOR -> (fragment as EditorDialogFragment).editedListener = {
                // TODO 该返回值是默认的，分情况处理返回值，下同
                typeEditPresenter.notifyUpdatingTypeName(it)
                true
            }
            TAG_TYPE_MARK_COLOR_PICKER -> (fragment as TypeMarkColorPickerDialogFragment).typeMarkColorPickedListener = {
                typeEditPresenter.notifyTypeMarkColorSelected(it)
                true
            }
            TAG_TYPE_MARK_PATTERN_PICKER -> (fragment as TypeMarkPatternPickerDialogFragment).typeMarkPatternPickedListener = {
                typeEditPresenter.notifyTypeMarkPatternSelected(it)
                true
            }
        }
    }

    override fun showInitialView(formattedType: FormattedType) {
        if (intent.getBooleanExtra(Constant.ENABLE_TRANSITION, false)) {
            window.requestFeature(Window.FEATURE_CONTENT_TRANSITIONS)
        }

        setContentView(R.layout.activity_type_edit)
        ButterKnife.bind(this)

        setSupportActionBar(toolbar)
        setupActionBar()

        onTypeNameChanged(formattedType.typeName, formattedType.firstChar)
        onTypeMarkColorChanged(formattedType.typeMarkColorInt, formattedType.typeMarkColorName!!)
        onTypeMarkPatternChanged(formattedType.hasTypeMarkPattern, formattedType.typeMarkPatternResId, formattedType.typeMarkPatternName)
    }

    override fun showTypeNameEditorDialog(originalEditorText: String) {
        EditorDialogFragment.Builder()
                .setTitle(R.string.title_dialog_type_name_editor)
                .setDefaultEditorText(originalEditorText)
                .setEditorHint(R.string.hint_type_name_editor_edit)
                .setEditButtonText(android.R.string.ok)
                .show(supportFragmentManager, TAG_TYPE_NAME_EDITOR)
    }

    override fun onTypeNameChanged(typeName: String, firstChar: String) {
        ic_type_mark.setInnerText(firstChar)
        text_type_name.text = typeName
        item_type_name.setDescriptionText(typeName)
    }

    override fun onTypeMarkColorChanged(colorInt: Int, colorName: String) {
        window.navigationBarColor = colorInt
        window.statusBarColor = colorInt
        toolbar.setBackgroundColor(ColorUtil.reduceSaturation(colorInt, 0.85f))
        ic_type_mark.setFillColor(colorInt)
        item_type_name.setThemeColor(colorInt)
        item_type_mark_color.setDescriptionText(colorName)
        item_type_mark_color.setThemeColor(colorInt)
        item_type_mark_pattern.setThemeColor(colorInt)
    }

    override fun onTypeMarkPatternChanged(hasPattern: Boolean, patternResId: Int, patternName: String?) {
        ic_type_mark.setInnerIcon(if (hasPattern) getDrawable(patternResId) else null)
        item_type_mark_pattern.setDescriptionText(if (hasPattern) patternName else getString(R.string.dscpt_unsettled))
    }

    override fun showTypeMarkColorPickerDialog(defaultColor: Int) {
        TypeMarkColorPickerDialogFragment.newInstance(defaultColor).show(supportFragmentManager, TAG_TYPE_MARK_COLOR_PICKER)
    }

    override fun showTypeMarkPatternPickerDialog(defaultPattern: String?) {
        TypeMarkPatternPickerDialogFragment.newInstance(defaultPattern).show(supportFragmentManager, TAG_TYPE_MARK_PATTERN_PICKER)
    }

    @OnClick(R.id.item_type_name, R.id.item_type_mark_color, R.id.item_type_mark_pattern)
    fun onClick(view: View) {
        when (view.id) {
            R.id.item_type_name -> typeEditPresenter.notifySettingTypeName()
            R.id.item_type_mark_color -> typeEditPresenter.notifySettingTypeMarkColor()
            R.id.item_type_mark_pattern -> typeEditPresenter.notifySettingTypeMarkPattern()
        }
    }
}
